docs/design/2025-03-17-sys-session-manage-enhancement.md

```diff
@@ -0,0 +1,167 @@
+
+With the newly implemented session pool that manages the session lifecycle, developers no longer need to worry about the session's state or lifecycle. 
+
+### Session
+
+The type `Session` is not a "real" one but a wrapper of the internal session, see:
+
+```go
+// Session is public to callers.
+type Session struct {
+    internal *session
+}
+```
+
+It provides a set of methods that proxy access to the internal session. When the caller invokes `WithSession`, a new `Session` instance is created and passed to the function. Once the function completes, the `Session` instance is destroyed, and only the internal session is returned to the pool.
+
+We have several reasons for proposing the above design:
+
+- The wrapped methods allow for additional checks to prevent misuse of the internal session. For example, if a caller attempts to invoke methods after the session has been returned to the pool, we can detect this and return an error.
+- The `Session` type can be extended with new methods, providing more convenient functionality for internal use.
+- The original session contains numerous methods, many of which may not be necessary for the caller. By filtering out unnecessary methods, we can expose only the required ones, simplifying the interface and improving usability.
+
+The internal session is not exposed to the caller and can only be accessed by the `Session` or `Pool` types. It has several fields, see:
+
+```go
+// session is private and can only be accessed by the `Pool` and `Session`.
+type session struct {
+    mu sync.Mutex
+    owner any
+    sctx sessionctx.Context
+    ...
+}
+```
+
+The `mu` protects the internal session from concurrent access, and the field `sctx` is the real session context that holds the session's state. 
+
+Let's provide further clarification on the `owner` field. The "owner" refers to the entity that holds the internal session, and only the owner is permitted to use it. When an internal session is in the pool, the pool itself is considered the owner. When a caller retrieves the session from the pool, the wrapped `Session` becomes the new owner, which is then passed to the caller.
+
+When the caller returns the session to the pool, ownership is first transferred back to the pool, which then places the session back into the pool. Since the public `Session` will not be reused, once it loses ownership, it can never regain it. This ensures that the `Session` becomes invalid once it has been used, preventing any further use after its lifecycle is completed.
+
+```go
+func (s *session) checkOwner(owner any) error {
+    s.mu.Lock()
+    defer s.mu.Unlock()
+    if s.owner == nil || s.owner != owner {
+        return errors.New("invalid owner")
+    }
+    return nil
+}
+
+func (s *Session) ExecuteInternal(ctx context.Context, sql string, args ...any) (sqlexec.RecordSet, error) {
+    if err := s.internal.checkOwner(s); err != nil {
```

Comment:
instead of `checkOwner`, I think we can add a atomic bool `closed` field to `*Session`. When `*Session` is put back to the pool, store `true` to `closed`.

In other words, lifecycle management belongs to `Session`, not `session`

Reply:
I think it is equivalent to the current design. When `session.owner != *Session` indicates the Session is closed, and when `session.owner == nil` means the session has been destroyed. One extra benefit for the current design is that the meta fields are more "private" and it's hard to construct a "illegal" `Session`. For example:

```
s1, err := pool.Get()
if err != nil {
    // handle error ...
}

s2 := *s1

// ...
```

For the current design, `s2.ExecuteInternal(...)` will produce an error because the inner session's owner is not s2. However, if `Session` keeps a `closed` field in it, this field will be cloned and both two session can be used at the same time

Reply:
The difference I want to stress is `session` should not care about upper layer's management like ownership checking. The upper layer's `Session` should done this only in its fields, without modifying `session` to add extra complexity.

For the new case of copying `Session`, I think it can be avoided by some nocopy mechanism.

Reply:
It is reasonable as long as the current design's semantics are self-consistent, I think. I've tried the above two practices and found each of them has the same complexity (both of them need to handle the meta fields, lock for thread-safe, etc...)

In this design, I think the semantics is clear that: 1. The pool cached `session` instead of `sessionctx.Context` to have more information about the internal context. 2. Each `session` has one and only one `owner` that has the privilege to operate on it. 3.  `Session` is only a simple wrapper to proxy the access and expose only necessary methods to callers.

Reply:
> same complexity

The difference is, your proposal is a two-direction relationship. `Session` need to manage `session`, and `session` also need to check `Session`. I'm not sure if it's really necessary to let `session` access `Session` in checking owner or other actions.

Reply:
You can think of `session` as the real `Session` in your design but we used `Session` in current design to provide methods to outside coders.

* `session` is used by `Pool` and `Session`. The public methods of `session` can be used by them.
* `Session` is used by the outside caller.